appengine/genetics/js/cage.js

```diff
@@ -0,0 +1,725 @@
+      proposingMouse.sex == askedMouse.sex) {
+    new Genetics.Cage.Event('MATE', proposingMouse.id, 'INCOMPATIBLE',
+        askedMouse.id).addToQueue();
+    return false;
+  }
+  // Check if other mouse is infertile.
+  if (askedMouse.fertility < 1) {
+    new Genetics.Cage.Event('MATE', proposingMouse.id, 'INFERTILE',
+        askedMouse.id).addToQueue();
+    return false;
+  }
+  // Ask second mouse whether it will mate with the proposing mouse.
+  var result = Genetics.Cage.runMouseFunction(askedMouse, 'mateAnswer',
+      proposingMouse);
+  // Check if function threw an error or caused a timeout.
+  if (result[0] == 'FAILURE') {
+    new Genetics.Cage.Event('MATE', proposingMouse.id, 'MATE_EXPLODED')
+        .addToQueue();
+    return false;
+  }
+  var response = result[1];
+  // Check if response is positive.
+  if (!response) {
+    new Genetics.Cage.Event('MATE', proposingMouse.id, 'REJECTION',
+        askedMouse.id).addToQueue();
+    return false;
+  }
+  return true;
+};
+
+/**
+ * Adds the mouse to the cage.
+ * @param {Genetics.Mouse} mouse The mouse to add to the cage.
+ */
+Genetics.Cage.born = function(mouse) {
+  // Adds child to population.
+  Genetics.Cage.aliveMice_.push(mouse);
+  Genetics.Cage.miceMap_[mouse.id] = mouse;
+  Genetics.Cage.life(mouse); // Queue life simulation for mouse.
+};
+
+/**
+ * Kills the mouse.
+ * @param {Genetics.Mouse} mouse The mouse to kill.
+ */
+Genetics.Cage.death = function(mouse) {
+  var index = Genetics.Cage.aliveMice_.indexOf(mouse);
+  if (index > -1) {
+    Genetics.Cage.aliveMice_.splice(index, 1);
+  }
+  delete Genetics.Cage.miceMap_[mouse.id];
+  clearTimeout(Genetics.Cage.lifePidsMap_[mouse.id]);
+  delete Genetics.Cage.lifePidsMap_[mouse.id];
+};
+
+/**
+ * Checks if mouse is alive.
+ * @param {Genetics.Mouse} mouse The mouse to check whether it is alive.
+ * @return {boolean} True if mouse is alive, false otherwise.
+ */
+Genetics.Cage.isAlive = function(mouse) {
+  return (Genetics.Cage.aliveMice_.indexOf(mouse) > -1);
+};
+
+/**
+ * Stops the game if end condition is met.
+ */
+Genetics.Cage.checkForEnd = function() {
+  // Check if there are no mice left.
+  if (Genetics.Cage.aliveMice_.length == 0) {
+    Genetics.Cage.end('NONE_LEFT');
+    return;
+  }
+  // Check if there is one mouse left.
+  else if (Genetics.Cage.aliveMice_.length == 1) {
+    var aliveMouse = Genetics.Cage.aliveMice_[0];
+    Genetics.Cage.end('ONE_LEFT', aliveMouse.pickFightOwner,
+        aliveMouse.chooseMateOwner, aliveMouse.mateAnswerOwner);
+    return;
+  }
+  // Check if there's no possible mate pairings left.
+  var potentialMate = null;
+  var matesExist = false;
+  for (var i = 0; i < Genetics.Cage.aliveMice_.length; i++) {
+    var mouse = Genetics.Cage.aliveMice_[i];
+    if (mouse.fertility > 0) {
+      if (!potentialMate) {
+        potentialMate = mouse;
+      }
+      else if (potentialMate.sex == Genetics.Mouse.Sex.HERMAPHRODITE ||
```

Comment:
Also I'm not sure if this is correct. Don't you need to check all possible pairs? Here, you fix potentialMate as the very first mouse which has positive fertility. But what if this particular mouse doesn't have a sex-compatible match among the later mice?

Also what if mouse.sex is hermaphrodite?